lib/mailbox.c

```diff
@@ -16,25 +16,122 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_debug.h>
+#include <rte_malloc.h>
+
 #include "gatekeeper_mailbox.h"
 
-/*
- * TODO Remove this preprocessing directive when there are
- * are calls to these functions; for now, the mailbox is
- * just an example of how to set up a library.
- */
-#if 0
+/* XXX Sample parameters, need to be tested for better performance. */
+#define GK_MEM_CACHE_SIZE (64)
+
+struct mailbox *
+create_mailbox(const char *tag,
+	int ele_count, int ele_size, unsigned int lcore_id)
+{
+	int ret;
+	char ring_name[128];
+	char pool_name[128];
+	char mailbox_name[128];
+	struct mailbox *mb = NULL;
+	unsigned int socket_id = rte_lcore_to_socket_id(lcore_id);
+
+	ret = snprintf(mailbox_name,
+		sizeof(mailbox_name), "%s_mailbox_%u", tag, lcore_id);
+	RTE_ASSERT(ret < sizeof(mailbox_name));
+
+	mb = (struct mailbox *)rte_calloc(mailbox_name, 1, sizeof(*mb), 0);
+	if (mb == NULL) {
+		RTE_LOG(ERR, MALLOC,
+			"mailbox: can't create mailbox %s (len = %d) at lcore %u!\n",
+			mailbox_name, ret, lcore_id);
+		return NULL;
+	}
+
+	ret = snprintf(ring_name,
+		sizeof(ring_name), "%s_mailbox_ring_%u", tag, lcore_id);
+	RTE_ASSERT(ret < sizeof(ring_name));
+
+	mb->ring = (struct rte_ring *)rte_ring_create(
+		ring_name, ele_count, socket_id, RING_F_SC_DEQ);
+    	if (mb->ring == NULL) {
+		RTE_LOG(ERR, RING,
+			"mailbox: can't create ring %s (len = %d) at lcore %u!\n",
+			ring_name, ret, lcore_id);
+		goto free_mailbox;
+	}
+
+	ret = snprintf(pool_name,
+		sizeof(pool_name), "%s_mailbox_pool_%d", tag, lcore_id);
+	RTE_ASSERT(ret < sizeof(pool_name));
+
+    	mb->pool = (struct rte_mempool *)rte_mempool_create(
+		pool_name, ele_count, ele_size, GK_MEM_CACHE_SIZE,
+		0, NULL, NULL, NULL, NULL, socket_id, 0);
+    	if (mb->pool == NULL) {
+		RTE_LOG(ERR, MEMPOOL,
+			"mailbox: can't create mempool %s (len = %d) at lcore %u!\n",
+			pool_name, ret, lcore_id);
+        	goto free_ring;
+    	}
+
+	return mb;
+
+free_ring:
+	rte_ring_free(mb->ring);
+free_mailbox:
+	rte_free(mb);
+
+	return NULL;
+}
+
+void *
+mb_get_entry(struct mailbox *mb)
+{
+	void *obj = NULL;
+	int ret = rte_mempool_get(mb->pool, &obj);
+	if (ret == -ENOENT) {
+		RTE_LOG(ERR, MEMPOOL,
+			"mailbox: not enough entries in the mempool.\n");
+		return NULL;
+	}
+
+	return obj;
+}
+
+int
+mb_put_entry(struct mailbox *mb, void *obj)
+{
+	int ret = rte_ring_mp_enqueue(mb->ring, obj);
+	if (ret == -EDQUOT) {
+		RTE_LOG(WARNING, RING,
+			"mailbox: quota exceeded. The objects have been enqueued, but the high water mark is exceeded.\n");
```

Comment:
Let's improve this warning with the following string:

`"mailbox: high water mark exceeded. The object has been enqueued.\n"`